### PR TITLE
Make sure we compact the LOH when aggressive GC is requested

### PR DESCRIPTION
**Pull request by cshung**

### CODE REVIEW

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -20839,6 +20839,11 @@ int gc_heap::joined_generation_to_condemn (BOOL should_evaluate_elevation,
         }
     }
 
+    if (settings.reason == reason_induced_aggressive)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:20842]: this should do `gc_data_global.gen_to_condemn_reasons.set_condition` like with the other reasons.

also now I'm looking at the code, it doesn't seem to set the condemn generation number anywhere for this mode? what is the behavior if you do do `GC.Collect (0, GCCollectionMode.Aggressive)?`? 0 is replaced by 2? if so I don't see this in the documentation. it should that this mode just does a full compacting GC.

**Review by Maoni0 — APPROVED** [review]

**Comment by MichalPetryka** [comment]
Isn't this a breaking change? The documentation currently says `By default, the LOH is not compacted.` and you're required to do `GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;` to compact it today.

**Comment by cshung** [comment]
> Isn't this a breaking change? The documentation currently says `By default, the LOH is not compacted.` and you're required to do `GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;` to compact it today.

It happens only when you specify a new enum value `Aggressive` for the `mode` parameter in `GC.Collect()`, so I believe it is an opt-in choice, not a breaking change of 'existing' behavior.

That being said, you sound concerned. Is there something that might break you if we do LOH compaction?

**Comment by MichalPetryka** [comment]
> Is there something that might break you if we do LOH compaction?

I've seen some legacy apps rely on this behaviour to have LOH arrays behave like POH ones, which is that they rely on them not being moved even when not pinned (but they're still keeping them alive). While relying on that is fragile, having such behaviour broken for them might cause some issues.

**Comment by cshung** [comment]
> > Is there something that might break you if we do LOH compaction?
> 
> I've seen some legacy apps rely on this behaviour to have LOH arrays behave like POH ones, which is that they rely on them not being moved even when not pinned (but they're still keeping them alive). While relying on that is fragile, having such behaviour broken might cause some issues for them.

But legacy app won't call `GC.Collect()` with the new collection mode, right? I am mostly counting on this mode being a new enum value to make sure it is safe to introduce new behavior.

**Comment by MichalPetryka** [comment]
Yeah if only `GC.Collect()` with a new mode will trigger this than I think it's okay (will need a bit of documentation changes though). I wasn't sure if the GC could trigger this aggressive behaviour in some other cases like low memory scenarios.

**Comment by cshung** [comment]
> Yeah if only `GC.Collect()` with a new mode will trigger this than I think it's okay (will need a bit of documentation changes though). I wasn't sure if the GC could trigger this aggressive behaviour in some other cases like low memory scenarios.

It won't. The LOH compaction behavior introduced in this change can happen only when `reason == reason_induced_aggressive`, and this can happen only through explicit `GC.Collect` API calls.

BTW, this is not the first time we perform LOH compaction on new options. Here are a few cases (already in the code base) that we will trigger LOH compaction because of "new" environment variables.

If `heap_hard_limit` is set and we decided LOH compaction is likely useful:
https://github.com/dotnet/runtime/blob/788f5726850d5bd85141f821f5329cb6e5a80a18/src/coreclr/gc/gc.cpp#L20792

If `GCConserveMem` is set and we found large fragmentation in LOH
https://github.com/dotnet/runtime/blob/788f5726850d5bd85141f821f5329cb6e5a80a18/src/coreclr/gc/gc.cpp#L20823

Suppose an app is truly legacy (e.g. you will not change it in anyway or run it in any other environments (e.g. container)), you might be safe to keep relying on the fact that LOH is not compacted unless requested. For example, are you really sure those legacy app will not be run on machine with the `GCConserveMem` environment variable set?

It is better safe than sorry. I would recommend us moving away from the legacy assumption that LOH won't be compacted and pin the object as appropriate.

**Pull request merged**